File: src/utils/reader.go
package utils

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"encoding/base64"

	"github.com/xuri/excelize/v2"
)

// ReadAccountsFromExcel reads account data from an Excel file and returns a slice of Account structs.
// It uses the provided start and end indices to determine which accounts to process.
//
// Usage example:
//
//	accounts, err := utils.ReadAccountsFromExcel("./data/accounts.xlsx", 1, 10)
//	if err != nil {
//	    log.Fatal(err)
//	}
//	for _, account := range accounts {
//	    fmt.Printf("Account: %+v\n", account)
//	}
//
// The function is thread-safe and will read all accounts until it finds an empty AUTH_TOKEN.
// It skips the header row and handles missing columns by setting empty strings.
func ReadAccountsFromExcel(filePath string, startIndex, endIndex int) ([]Account, error) {
	// Lock the mutex before accessing the file
	ExcelMutex.Lock()
	defer ExcelMutex.Unlock()

	// Open the Excel file
	f, err := excelize.OpenFile(filePath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	// Get the first sheet
	sheetName := f.GetSheetName(0)

	// Get all rows from the sheet
	rows, err := f.GetRows(sheetName)
	if err != nil {
		return nil, err
	}

	var accounts []Account

	// Skip the first row (header) and process remaining rows
	for i := 1; i < len(rows); i++ {
		row := rows[i]

		// Check if we have enough columns
		if len(row) < 1 {
			continue
		}

		// Stop if AUTH_TOKEN is empty
		if row[0] == "" {
			break
		}

		// Create account struct, handling possible missing columns
		account := Account{
			AuthToken: row[0],
			Proxy:     "",
			Username:  "",
			Status:    "",
		}

		// Safely assign other fields if they exist
		if len(row) > 1 {
			account.Proxy = row[1]
		}
		if len(row) > 2 {
			account.Username = row[2]
		}
		if len(row) > 3 {
			account.Status = row[3]
		}

		accounts = append(accounts, account)
	}

	// If both are 1 and 0 or equal, process all accounts
	if (startIndex == 1 && endIndex == 0) || startIndex == endIndex {
		return accounts, nil
	}

	// Validate indices
	if startIndex < 1 {
		startIndex = 1
	}
	if endIndex == 0 || endIndex > len(accounts) {
		endIndex = len(accounts)
	}

	// Return the slice of accounts within the range
	return accounts[startIndex-1 : endIndex], nil
}

func UserInputToSlice(textToShow string) []string {
	reader := bufio.NewReader(os.Stdin)

	fmt.Print("\n" + textToShow + ": ")
	userInput, _ := reader.ReadString('\n')
	userInput = strings.TrimSpace(userInput)
	userInput = strings.ReplaceAll(userInput, ",", " ")

	return strings.Fields(userInput)
}

func UserInputInteger(textToShow string) (int, bool) {
	var input int

	fmt.Print(textToShow + ": ")
	reader := bufio.NewReader(os.Stdin)
	userInput, _ := reader.ReadString('\n')
	userInput = strings.TrimSpace(userInput)

	input, err := strconv.Atoi(userInput)
	if err != nil {
		Logger{}.Error("Wrong input.")
		return input, false
	}
	return input, true
}

func EncodeFileToBase64(filePath string) (string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return "", fmt.Errorf("failed to read file: %w", err)
	}
	return base64.StdEncoding.EncodeToString(data), nil
}

func ReadTxtFile(fileName string, filePath string) []string {
	file, err := os.Open(filePath)
	if err != nil {
		log.Fatalf("Failed to open file: %v", err)
		return []string{}
	}
	defer func(file *os.File) {
		err := file.Close()
		if err != nil {
			panic(err)
		}
	}(file)

	var items []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		items = append(items, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("Failed to read file: %v", err)
		return []string{}
	}

	Logger{}.Info("Successfully loaded %d %s.", len(items), fileName)
	return items
}

func ReadPictures(path string) ([]string, bool) {
	var encodedImages []string

	files, err := os.ReadDir(path)
	if err != nil {
		Logger{}.Error("Failed to read pictures dir: %s", err)
		return encodedImages, false
	}

	for _, file := range files {
		if strings.HasSuffix(file.Name(), ".png") || strings.HasSuffix(file.Name(), ".jpg") || strings.HasSuffix(file.Name(), ".jpeg") {
			filePath := filepath.Join(path, file.Name())

			imageFile, err := os.ReadFile(filePath)
			if err != nil {
				Logger{}.Error("Failed to read the picture: %s", err)
			}

			encodedImage := base64.StdEncoding.EncodeToString(imageFile)
			encodedImages = append(encodedImages, encodedImage)
		}
	}

	return encodedImages, true
}

File: src/utils/reader.py
import base64
import json
import os
from typing import List
from loguru import logger
import threading
import pandas as pd
from .constants import Account
from .proxy_parser import Proxy


def read_txt_file(file_name: str, file_path: str) -> list:
    with open(file_path, "r") as file:
        items = [line.strip() for line in file]

    logger.success(f"Successfully loaded {len(items)} {file_name}.")
    return items


def split_list(lst, chunk_size=90):
    return [lst[i : i + chunk_size] for i in range(0, len(lst), chunk_size)]


# Mutex for thread-safe Excel file access
excel_mutex = threading.Lock()


def read_accounts_from_excel(
    file_path: str, start_index: int = 1, end_index: int = 0
) -> list:
    """
    Reads account data from an Excel file and returns a list of Account objects.
    It uses the provided start and end indices to determine which accounts to process.

    Args:
        file_path: Path to the Excel file
        start_index: Starting index (1-based, default is 1)
        end_index: Ending index (0 means all accounts, default is 0)

    Returns:
        List of Account objects

    Example:
        accounts = read_accounts_from_excel("./data/accounts.xlsx", 1, 10)
        for account in accounts:
            print(f"Account: {account}")
    """
    # Lock the mutex before accessing the file
    with excel_mutex:
        try:
            # Check if file exists
            if not os.path.exists(file_path):
                raise FileNotFoundError(f"Excel file not found: {file_path}")

            # Read Excel file
            df = pd.read_excel(file_path)

            # Check if required columns exist
            if "AUTH_TOKEN" not in df.columns:
                raise ValueError("Excel file must have an 'AUTH_TOKEN' column")

            accounts = []

            # Process rows until empty AUTH_TOKEN
            for index, row in df.iterrows():
                # Skip rows with empty auth_token
                if (
                    pd.isna(row.get("AUTH_TOKEN", ""))
                    or row.get("AUTH_TOKEN", "") == ""
                ):
                    break

                # Get auth token
                auth_token = row.get("AUTH_TOKEN", "")

                # Get proxy with default empty string
                proxy_str = row.get("PROXY", "")
                if pd.isna(proxy_str):
                    proxy_str = ""

                # Process non-empty proxy through proxy_parser
                if proxy_str:
                    try:
                        proxy_obj = Proxy.from_str(proxy_str)
                        # Convert to user:pass@ip:port format
                        proxy_str = proxy_obj.get_default_format()
                    except Exception as e:
                        # Add row number to error message
                        raise ValueError(
                            f"Failed to parse proxy '{proxy_str}' on row {index + 2}: {str(e)}"
                        )

                # Create account object
                account = Account(auth_token=auth_token, proxy=proxy_str)
                accounts.append(account)

            if not accounts:
                raise ValueError("No valid accounts found in the Excel file. Please check that the file has accounts with valid AUTH_TOKEN values.")

            # If both are 1 and 0 or equal, process all accounts
            if (start_index == 1 and end_index == 0) or start_index == end_index:
                filtered_accounts = accounts
            else:
                # Validate indices
                if start_index < 1:
                    start_index = 1
                if end_index == 0 or end_index > len(accounts):
                    end_index = len(accounts)

                # Return accounts within the range
                filtered_accounts = accounts[start_index - 1 : end_index]

            logger.success(
                f"Successfully loaded {len(filtered_accounts)} accounts from Excel file."
            )
            return filtered_accounts

        except pd.errors.EmptyDataError:
            logger.error("The Excel file is empty")
            return []
        except Exception as e:
            logger.error(f"Error reading Excel file: {e}")
            return []


async def read_pictures(file_path: str) -> List[str]:
    """
    Считывает изображения из указанной папки и кодирует их в base64

    Args:
        file_path: Путь к папке с изображениями

    Returns:
        Список закодированных изображений в формате base64
    """
    encoded_images = []

    # Создаем папку, если она не существует
    os.makedirs(file_path, exist_ok=True)
    logger.info(f"Reading pictures from {file_path}")

    try:
        # Получаем список файлов
        files = os.listdir(file_path)

        if not files:
            logger.warning(f"No files found in {file_path}")
            return encoded_images

        # Обрабатываем каждый файл
        for filename in files:
            if filename.endswith((".png", ".jpg", ".jpeg")):
                # Формируем полный путь к файлу
                image_path = os.path.join(file_path, filename)

                try:
                    with open(image_path, "rb") as image_file:
                        encoded_image = base64.b64encode(image_file.read()).decode(
                            "utf-8"
                        )
                        encoded_images.append(encoded_image)
                except Exception as e:
                    logger.error(f"Error loading image {filename}: {str(e)}")

    except FileNotFoundError:
        logger.error(f"Directory not found: {file_path}")
    except PermissionError:
        logger.error(f"Permission denied when accessing: {file_path}")
    except Exception as e:
        logger.error(f"Error reading pictures from {file_path}: {str(e)}")

    return encoded_images
